src/rules/no-extra-reactive-curlies.ts

```diff
@@ -0,0 +1,60 @@
+import type { TSESTree } from "@typescript-eslint/types"
+import { createRule } from "../utils"
+
+export default createRule("no-extra-reactive-curlies", {
+  meta: {
+    docs: {
+      description:
+        "disallow wrapping single reactive statements in curly braces",
+      category: "Stylistic Issues",
+      recommended: false,
+      conflictWithPrettier: false,
+    },
+    fixable: "code",
+    schema: [],
+    messages: {
+      extraCurlies: `Do not wrap reactive statements in curly braces unless necessary.`,
+    },
+    type: "suggestion",
+  },
+  create(context) {
+    return {
+      [`SvelteReactiveStatement > BlockStatement[body.length=1]`]: (
+        node: TSESTree.BlockStatement,
+      ) => {
+        // $: { foo = "bar"; }
+        // Only want to transform if the contents of the block is a single assignment
+        // Anything else gets us into potentially weird territory and probably isn't worth handling
+        if (
+          node.body[0].type !== "ExpressionStatement" ||
+          node.body[0].expression.type !== "AssignmentExpression"
```

Comment:
I would like to omit the curly braces in the reactive statement as follows, not just in the assignment:

```svelte
<script>
$: {
  if (foo) { // and `for`, `switch`, `while`, `try`, and more
    // ...
  }
}
$: {
  fn(foo, bar)
}
</script>
```

Do you think these should be covered by this rule?

Reply:
Ideally yes! The more I think about this the more I think the "invalid with braces removed" examples you provided are uncommon enough that we shouldn't try to account for them in this rule and let folks disable it if they want to do something as weird as defining a class in a reactive statement.

Otherwise no matter what it's going to end up being a chunk of conditional logic that is always gonna be hard to reason about in the rule, and I'm worried that either the whitelist or blacklist approach are both gonna be a maintenance hassle. Thoughts?

Reply:
I think it probably makes sense to use an allowed list approach.
The spec seems to allow statements. (However, since it is different from ESTree, the declaration is not included.)
https://tc39.es/ecma262/2022/multipage/ecmascript-language-statements-and-declarations.html#prod-LabelledItem

Given the specs and the actual method used, I think the complete list is below.

- BlockStatement
- ExpressionStatement
- IfStatement
- ForStatement
- ForInStatement
- ForOfStatement
- WhileStatement
- DoWhileStatement
- SwitchStatement
- TryStatement

The following is a code example:

```js
$: {} // BlockStatement
/* Defined as VariableStatement in ECMAScript. */
$: var a = "" // VariableDeclaration with var (not let and const)
$: ; // EmptyStatement
$: foo() // ExpressionStatement
$: if (foo) {} // IfStatement
/* Defined as BreakableStatement in ECMAScript. */
$: for(;;) {} // ForStatement
$: for(const a in lint) {} // ForInStatement
$: for(const a of lint) {} // ForOfStatement
$: while(foo) {} // WhileStatement
$: do {} while(foo) // DoWhileStatement
$: switch (a) { case 1: {} } // SwitchStatement
// $: continue // ContinueStatement (Syntax Error: not available at the top level.)
// $: break // BreakStatement (Syntax Error: not available at the top level.)
$: return // ReturnStatement ( globalReturn: true )
// $: with (foo) {} // WithStatement (Syntax Error: not available in strict mode.)
$: throw new Error() // ThrowStatement
$: try {} finally {} // TryStatement


/* Defined as LexicalDeclaration in ECMAScript. */
// $: const b = "" // Syntax Error
// $: let c = "" // Syntax Error
/* Defined as ClassDeclaration in ECMAScript. */
// $: class MyClass {} // Syntax Error
/* Defined as HoistableDeclaration in ECMAScript. */
// $: function fn(){} // Syntax Error (strict mode)
```

Reply:
I forgot to include the LabeledStatement 😓 

```js
$: foo: for (;;) {} // LabeledStatement
```

Reply:
I can look into supporting all of the various possibilities but I'm not particularly excited about it. There's more that are allowed than disallowed, so I think a blacklist makes more sense. But even paring it down by doing a blacklist it's hard to justify supporting stuff like not transforming a reactive class declaration (wouldn't that throw an error about class redeclaration?) or a reactive`const`? The only way that makes any sense is with more than one statement because you can't access the `const` outside of the reactive statement.

I'm not really sure where to take this rule because either approach feels like a massive explosion of complexity for very small returns. Maybe instead of a fixable rule it just offers a suggestion and we keep it simpler?

Reply:
> Maybe instead of a fixable rule it just offers a suggestion and we keep it simpler?

Yeah. I think that is a good idea :+1: